core/state/snapshot/difflayer.go

```diff
@@ -0,0 +1,337 @@
+	switch parent := dl.parent.(type) {
+	case *diskLayer:
+		return parent.number, dl.number
+	case *diffLayer:
+		dl.lock.Lock()
+		defer dl.lock.Unlock()
+
+		dl.parent = parent.flatten()
+		if dl.parent.(*diffLayer).memory < memory {
+			diskNumber, _ := parent.parent.Info()
+			return diskNumber, parent.number
+		}
+	default:
+		panic(fmt.Sprintf("unknown data layer: %T", parent))
+	}
+	// If the bottommost layer is larger than our memory cap, persist to disk
+	var (
+		parent = dl.parent.(*diffLayer)
+		base   = parent.parent.(*diskLayer)
+		batch  = base.db.NewBatch()
+	)
+	parent.lock.RLock()
+	defer parent.lock.RUnlock()
+
+	// Start by temporarilly deleting the current snapshot block marker. This
+	// ensures that in the case of a crash, the entire snapshot is invalidated.
+	rawdb.DeleteSnapshotBlock(batch)
+
+	// Push all the accounts into the database
+	for hash, data := range parent.accountData {
+		if len(data) > 0 {
+			// Account was updated, push to disk
+			rawdb.WriteAccountSnapshot(batch, hash, data)
+			base.cache.Set(string(hash[:]), data)
+
+			if batch.ValueSize() > ethdb.IdealBatchSize {
+				if err := batch.Write(); err != nil {
+					log.Crit("Failed to write account snapshot", "err", err)
+				}
+				batch.Reset()
+			}
+		} else {
+			// Account was deleted, remove all storage slots too
+			rawdb.DeleteAccountSnapshot(batch, hash)
+			base.cache.Set(string(hash[:]), nil)
+
+			it := rawdb.IterateStorageSnapshots(base.db, hash)
+			for it.Next() {
+				if key := it.Key(); len(key) == 65 { // TODO(karalabe): Yuck, we should move this into the iterator
+					batch.Delete(key)
+					base.cache.Delete(string(key[1:]))
+				}
+			}
+			it.Release()
+		}
+	}
+	// Push all the storage slots into the database
+	for accountHash, storage := range parent.storageData {
+		for storageHash, data := range storage {
+			if len(data) > 0 {
+				rawdb.WriteStorageSnapshot(batch, accountHash, storageHash, data)
+				base.cache.Set(string(append(accountHash[:], storageHash[:]...)), data)
+			} else {
+				rawdb.DeleteStorageSnapshot(batch, accountHash, storageHash)
+				base.cache.Set(string(append(accountHash[:], storageHash[:]...)), nil)
+			}
+		}
+		if batch.ValueSize() > ethdb.IdealBatchSize {
+			if err := batch.Write(); err != nil {
+				log.Crit("Failed to write storage snapshot", "err", err)
+			}
+			batch.Reset()
+		}
+	}
+	// Update the snapshot block marker and write any remainder data
+	base.number, base.root = parent.number, parent.root
+
+	rawdb.WriteSnapshotBlock(batch, base.number, base.root)
+	if err := batch.Write(); err != nil {
+		log.Crit("Failed to write leftover snapshot", "err", err)
+	}
+	dl.parent = base
+
+	return base.number, dl.number
+}
+
+// flatten pushes all data from this point downwards, flattening everything into
+// a single diff at the bottom. Since usually the lowermost diff is the largest,
+// the flattening bulds up from there in reverse.
+func (dl *diffLayer) flatten() snapshot {
```

Comment:
We need to generate a new `snapshot` instead of reusing the parent struct (we can reuse the data fields) so that we can invalidate the parent (anyone holding a reference to it should stop working) without invalidating the new flattened thing.

Reply:
We also need support to invalidate old references to the disk layer!

Reply:
> We need to generate a new snapshot instead of reusing the parent struct 

Another way to solve it is to have parents maintain a mapping to `children`, so the parent can set a flag on the `child` whenever it becomes modified/flattened by another child.

Reply:
I implemented a fourth option in https://github.com/karalabe/go-ethereum/pull/16 . The root hash of the parent is not required, since the only thing that can ever happen is that more blocks are pushed into it. Thus, it's sufficient to track the parent `number`, which we can access by `atomic`s, and not have to obtain lock every time we search for an item.